src/announcements/react.js

```diff
@@ -0,0 +1,249 @@
+// @flow
+import uniq from "lodash/uniq";
+import intersection from "lodash/intersection";
+import React, {
+  createContext,
+  useContext,
+  useEffect,
+  useMemo,
+  useReducer,
+} from "react";
+import type { Announcement } from "./types";
+import { fetchAnnouncements } from "./logic";
+
+type Context = {
+  language: string,
+  currencies: string[],
+  date: Date,
+};
+
+type Props = {
+  children: React$Node,
+  onLoad: () => Promise<{
+    announcements: Announcement[],
+    seenIds: string[],
+  }>,
+  onSave: ({
+    announcements: Announcement[],
+    seenIds: string[],
+  }) => Promise<void>,
+  context: Context,
+};
+
+type Cache = {
+  [id: string]: Announcement,
+};
+
+type State = {
+  seenIds: string[],
+  allIds: string[],
+  cache: Cache,
+  isLoading: boolean,
+  error: ?Error,
+};
+
+type API = {
+  updateCache: () => Promise<void>,
+  setAsSeen: (seenIds: string[]) => void,
+};
+
+type AnnouncementContext = State & API;
+
+type SetAsSeenAction = { type: "setAsSeen", seenIds: string[] };
+
+type UpdateCacheSuccessAction = {
+  type: "updateCacheSuccess",
+  announcements: Announcement[],
+};
+
+type UpdateCacheErrorAction = {
+  type: "updateCacheError",
+  error: Error,
+};
+
+type UpdateCachePendingAction = {
+  type: "updateCachePending",
+};
+
+type LoadCacheAction = {
+  type: "updateCachePending",
+  announcements: Announcement[],
+  seenIds: string[],
+};
+
+type Action =
+  | SetAsSeenAction
+  | UpdateCacheSuccessAction
+  | UpdateCacheErrorAction
+  | UpdateCachePendingAction
+  | LoadCacheAction;
+
+const AnnoucementsContext = createContext<AnnouncementContext>({});
+
+const initialState: State = {
+  cache: {},
+  seenIds: [],
+  allIds: [],
+  error: null,
+  isLoading: false,
+};
+const reducer = (state: State, action: Action) => {
+  switch (action.type) {
+    case "setAsSeen":
+      return {
+        ...state,
+        seenIds: uniq([...state.seenIds, ...action.seenIds]),
+      };
+
+    case "loadCache": {
+      const cache = action.announcements.reduce((acc, curr) => {
+        acc[curr.uuid] = curr;
+        return acc;
+      }, {});
```

Comment:
reduce to build a `{}` is imo an anti-pattern (fact we mutate the accumulator), we could just forEach and with a `const cache = {};` that we fill from inside forEach.